pxtcompiler/emitter/languageservice.ts

```diff
@@ -0,0 +1,650 @@
+    const COMPLETION_IN_SCOPE_VAR_WEIGHT = 5;
+    const COMPLETION_MATCHING_PARAM_TYPE_WEIGHT = 10;
+
+    // TODO(dz):
+    // interface 
+    // class
+    // function and return
+
+    export function getCallSymbol(callExp: CallExpression): SymbolInfo {// pxt symbol
+        const qName = callExp?.pxt?.callInfo?.qName
+        const api = lastApiInfo.apis.byQName[qName]
+        return api
+    }
+
+    export function getParameterTsType(callSym: SymbolInfo, paramIdx: number, blocksInfo: BlocksInfo): string | undefined {
+        if (!callSym || paramIdx < 0)
+            return undefined;
+
+        const paramDesc = callSym.parameters[paramIdx]
+        let result = paramDesc.type;
+
+        // check if this parameter has a shadow block, if so use the type from that instead
+        if (callSym.attributes._def) {
+            const blockParams = callSym.attributes._def.parameters
+            const blockParam = blockParams[paramIdx]
+
+            const shadowId = blockParam.shadowBlockId
+            if (shadowId) {
+                const shadowBlk = blocksInfo.blocksById[shadowId]
+                const shadowApi = lastApiInfo.apis.byQName[shadowBlk.qName]
+
+                const isPassThrough = shadowApi.attributes.shim === "TD_ID"
+                if (isPassThrough && shadowApi.parameters.length === 1) {
+                    const realTyp = shadowApi.parameters[0].type
+                    result = realTyp
+                }
+            }
+        }
+
+        return result
+    }
+
+    export function getApisForTsType(pxtType: string, location: Node, tc: TypeChecker, symbols: CompletionSymbol[]): CompletionSymbol[] {
+        // any apis that return this type?
+        // TODO: if this becomes expensive, this can be cached between calls since the same
+        // return type is likely to occur over and over.
+        const apisByRetType: pxt.Map<CompletionSymbol[]> = {}
+        symbols.forEach(i => {
+            apisByRetType[i.symbol.retType] = [...(apisByRetType[i.symbol.retType] || []), i]
+        })
+
+        const retApis = apisByRetType[pxtType]
+
+        // any enum members?
+        let enumVals: SymbolInfo[] = []
+        for (let r of retApis) {
+            const asTsEnum = getTsSymbolFromPxtSymbol(r.symbol, location, SymbolFlags.Enum)
+            if (asTsEnum) {
+                const enumType = tc.getTypeOfSymbolAtLocation(asTsEnum, location)
+                const mems = getEnumMembers(tc, enumType)
+                const enumValQNames = mems.map(e => enumMemberToQName(tc, e))
+                const symbols = enumValQNames.map(n => lastApiInfo.apis.byQName[n])
+                enumVals = [...enumVals, ...symbols]
+            }
+        }
+
+        return [...retApis, ...completionSymbols(enumVals, COMPLETION_DEFAULT_WEIGHT)]
+    }
+
+
+    export function getBasicKindDefault(kind: SyntaxKind, isPython: boolean): string | undefined {
+        switch (kind) {
+            case SK.StringKeyword: return "\"\"";
```

Comment:
This used to be:
```
case SK.StringKeyword: return "";
```
which caused no default value to be emitted for string parameters. This was found by the new test suite.